src/math/exp_fcn.c

```diff
@@ -0,0 +1,177 @@
+ *		  int64_t in_1
+ *		  uint64_t *ptroutbitshi
+ *		  uint64_t *ptroutbitslo
+ * Multiplication of two signed int64 numbers
+ * Return Type	: void
+ * Description:Perform element-wise multiplication on in_0 and in_1
+ * while keeping the required product word length and fractional
+ * length in mind. mul_s64 function divide the 64-bit quantities
+ * into two 32-bit words, multiply the low words to produce the
+ * lowest and second-lowest words in the result, then both pairs
+ * of low and high words from different numbers to produce the
+ * second and third lowest words in the result, and finally both
+ * high words to produce the two highest words in the outcome.
+ * Add them all up, taking carry into consideration.
+ *
+ * The 64 x 64 bit multiplication of operands in_0 and in_1 is
+ * shown in the image below. The 64-bit operand in_0, in_1 is
+ * represented by the notation in0_H, in1_H for the top 32 bits
+ * and in0_L, in1_L for the bottom 32 bits.
+ *
+ *				in0_H : in0_L
+ *			x	in1_H : in1_L
+ *			---------------------
+ *	P0			in0_L x in1_L
+ *	P1		in0_H x in1_L		64 bit inner multiplication
+ *	P2		in0_L x in1_H		64 bit inner multiplication
+ *	P3	in0_H x in1_H
+ *			--------------------
+ *			[64 x 64 bit multiplication] sum of inner products
+ * All combinations are multiplied by one another and then added.
+ * Each inner product is moved into its proper power location, given the names
+ * of the inner products, redoing the addition where 000 represents 32 zero
+ * bits. The inner products can be added together in 64 bit addition. The sum
+ * of two 64-bit numbers yields a 65-bit output.
+ *		   (P0H:P0L)
+ *		P1H(P1L:000)
+ *		P2H(P2L:000)
+ *	P3H:P3L(000:000)
+ *	.......(aaa:P0L)
+ * By combining P0H:P0L and P1L:000. This can lead to a carry, denote as CRY0.
+ * The partial result is then multiplied by P2L:000.
+ * We call it CRY1 because it has the potential to carry again.
+ *	(CRY0 + CRY1)P0H:P0L
+ *	(	 P1H)P1L:000
+ *	(	 P2H)P2L:000
+ *	(P3H:	 P3L)000:000
+ *	--------------------
+ *	(ccc:bbb)aaa:P0L
+ * P1H, P2H, and P3H:P3L are added to the carry CRY0 + CRY1. This increase will
+ * not result in an overflow.
+ *
+ */
+static inline void mul_s64(int64_t in_0, int64_t in_1, uint64_t *ptroutbitshi,
+				uint64_t *ptroutbitslo)
+{
+	uint64_t absin0 = ABS(in_0);
+	uint64_t absin1 = ABS(in_1);
+	uint64_t in0hi = absin0 >> 32;
+	uint64_t in0lo = absin0 & UINT32_MAX;
+	uint64_t in1hi = absin1 >> 32;
+	uint64_t prodhilo;
+	uint64_t prodlohi;
+
+	absin0 = absin1 & UINT32_MAX;
+	/* multiplication */
+	prodhilo = in0hi * absin0;
+	prodlohi = in0lo * in1hi;
+	absin0 *= in0lo;
+	absin1 = 0;
+	in0lo = absin0 + (prodlohi << 32);
+	absin1 = in0lo < absin0 ? 1 : 0;
+	absin0 = in0lo;
+	/* shift and add */
+	in0lo += prodhilo << 32;
+	if (in0lo < absin0)
+		absin1++;
+	/* carry */
+	absin0 = absin1 + in0hi * in1hi + (prodlohi >> 32) + (prodhilo >> 32);
+	/* 2's complement */
+	if (in_0  && (in_1  && ((in_0 > 0) != (in_1 > 0)))) {
```

Comment:
some of these additional brackets are unneeded and just make it harder to read

Reply:
I see errors
```
error: suggest parentheses around comparison in operand of ‘!=’ [-Werror=parentheses]
   98 |         if (in_0  && (in_1  && (in_0 > 0 != in_1 > 0))) {
      |                                 ~~~~~^~~
```

Reply:
> error: suggest parentheses around comparison in operand of ‘!=’ [-Werror=parentheses]
>   98 |         if (in_0  && (in_1  && (in_0 > 0 != in_1 > 0))) {

You tried to remove the only parentheses that seem useful. I don't want to be guessing the priority between `>` and `!=`. It should be possible to remove the other parentheses without issue, hopefully like this:
```suggestion
	if (in_0  && in_1  && (in_0 > 0) != (in_1 > 0)) {
```